tempesta_fw/http.c

```diff
@@ -2641,128 +2780,346 @@ tfw_http_hm_drop_resp(TfwHttpResp *resp)
 	tfw_http_msg_free((TfwHttpMsg *)req);
 }
 
+static void
+tfw_http_skb_queue(TfwHttpMsg *hm, struct sk_buff *skb,
+		   tfw_http_parse_stage_t stage)
+{
+	if (!tfw_http_msg_is_streamed(hm)) {
+		ss_skb_queue_tail(&hm->msg.head_skb, skb);
+		return;
+	}
+
+	spin_lock(&hm->msg.stream_lock);
+
+	switch (stage)
+	{
+	case TFW_HTTP_PARSE_HEADERS:
+		ss_skb_queue_tail(&hm->msg.head_skb, skb);
+		break;
+	case TFW_HTTP_PARSE_BODY:
+		ss_skb_queue_tail(&hm->msg.body_skb, skb);
+		break;
+	case TFW_HTTP_PARSE_TRAILER:
+		ss_skb_queue_tail(&hm->msg.trailer_skb, skb);
+		break;
+	default:
+		BUG();
+	}
+
+	spin_unlock(&hm->msg.stream_lock);
+}
+
+static int
+tfw_http_stream_req(TfwHttpReq *req)
+{
+	TfwSrvConn *conn = (TfwSrvConn *)req->state.stream_conn;
+	int r = 0;
+
+	TFW_DBG2("%s: resp=[%p]\n", __func__, req);
+
+	spin_lock(&conn->fwd_qlock);
+
+	if (!conn->msg_sent || ((TfwHttpReq *)conn->msg_sent != req))
+		goto err;
+	if (!tfw_srv_conn_get_if_live(conn))
+		goto err;
+	tfw_cli_conn_mod_katimer((TfwCliConn *)req->conn);
+	r = __tfw_http_conn_stream(req->state.stream_conn, (TfwHttpMsg *)req);
+	tfw_srv_conn_put(conn);
+err:
+	spin_unlock(&conn->fwd_qlock);
+
+	return r;
+}
+
+/**
+ * Parse and add a new skb into current message.
+ *
+ * Before the call @skb is not attached to the message. If parsing succeed,
+ * @skb is either attached to the message and @skb pointer is set to NULL;
+ * or it's split into two parts, first is attached to the message, second is
+ * returned in @skb pointer.
+ */
+static int
+tfw_http_parse_skb(TfwConn *conn, struct sk_buff **skb, unsigned int *off,
+		   TfwFsmData *data_up)
+{
+	TfwHttpMsg *hm = (TfwHttpMsg *)conn->msg;
+	TfwHttpParser *parser = &hm->parser;
+	bool client = TFW_CONN_TYPE(conn) & Conn_Clnt;
+	ss_skb_actor_t actor = client ? tfw_http_parse_req : tfw_http_parse_resp;
+	tfw_http_parse_stage_t stage;
+	unsigned int data_off;
+	size_t to_read;
+	int r = TFW_BLOCK;
+
+repeat:
+	if (likely(tfw_cfg_proxy_buff_sz > hm->msg.len))
+		to_read = tfw_cfg_proxy_buff_sz - hm->msg.len;
+	else
+		to_read = ULONG_MAX;
+
+	/*
+	 * Process/parse data in the SKB.
+	 * @data_off points at the start of data for processing.
+	 * @off is the current offset of data to process in
+	 * the SKB. After processing @off points at the end
+	 * of latest data chunk. However processing may have
+	 * stopped in the middle of the chunk. Adjust it to point
+	 * to the right location within the chunk.
+	 */
+	stage = tfw_http_parse_stage(hm);
+	parser->skb = *skb;
+	data_off = *off;
+	r = ss_skb_process(*skb, off, to_read, actor, hm);
```

Comment:
The buffering significantly affects TCP behaviour, so please develop a functional TCP test which analyzes TCP flow (ACKs and receive window announcements) as well as changings in TCP receive and send buffers for both the server and client sides. Also check drop counters for the sockets (add them if necessary, but I reckon the stack already accounts them).

Reply:
It seems the test isn't done yet...